### ISSUE TITLE
Add support for SSL

### ISSUE THREAD

**Issue by pchiusano** [functional-streams-for-scala/fs2#102]
This is a feature request. We'd like to implement the following combinator, or something like it:

``` Scala
def ssl(config: SSLParameters)(e: Exchange[Bytes, Bytes]): Exchange[Bytes,Bytes]
```

Where `config` is perhaps of type [`SSLParameters`](http://docs.oracle.com/javase/7/docs/api/javax/net/ssl/SSLParameters.html), or whatever type encapsulates all the options one might use in configuring an SSL connection.

I'd like it to be done as an `Exchange` transformer so this can be easily used with the `nio` client and servers that have already been written (as well as with ordinary blocking `java.io` sockets).

The returned `Exchange` should block on completion of an SSL handshake before sending or receiving any bytes, and after the handshake is complete, bytes should be encrypted using the negotiated encrypted channel.

It seems like one could use [`SSLEngine`](http://docs.oracle.com/javase/7/docs/api/javax/net/ssl/SSLEngine.html) directly for this. It's a pretty awful API, which is why wrapping it in something sane would be valuable. I am open to borrowing or stealing code from other libraries to help with this task, but would strongly prefer not to include any other dependencies in scalaz-stream.

If you're interested in contributing and have questions about the requirements or how to approach this, please feel free to ask questions in the comments.

**Comment by channingwalton** [comment]
hi all,
whats happening with this? It would be very useful for us if it were available.

Channing

**Comment by djspiewak** [comment]
@channingwalton No one is working on this at present.  It would be really great, but we have some more core priorities at present.

**Comment by channingwalton** [comment]
Understood. This would be very useful for us so I'll stare at it very hard. It would be good to know what @dscleaver's work is missing if anything.

**Comment by eamelink** [comment]
I'm interested in looking at this. So far I've experimented a bit with hacking an SSLEngine into `fs2.io.tcp.Socket`, and got as far as succesfully handshaking and running an echoserver over TLS.

I've learned a lot about SSLEngine doing this :)

With `Exchange` gone, the question of where to put it is back on the table I think.

Two options I considered are:

* Doing it at `Stream` level: creating a pair of `Pipe[F, Byte, Byte]`, one for encoding the outgoing stream, and one for decoding the incoming stream.
* Doing it at `Socket` level.

##### Ultrashort SSLEngine intro

SSLEngine has two methods: `wrap` and `unwrap`. They both take two `ByteBuffer`: one for the application data, and one for the net data. `wrap` wraps the application data in an SSL packet and puts it in the net data buffer, and `unwrap` vice versa. 

`wrap` and `unwrap` can be used concurrently, so during normal operation it's smooth sailing.

`wrap` and `unwrap` return a handshake status, the most important are `OK`, `NEED_WRAP` and `NEED_UNWRAP`. Each of these statuses can come back from each call to `wrap` and `unwrap`. These extra invocations of `wrap` and `unwrap`, during which `SSLEngine` does not consume nor produce application data, give it an opportunity to send and receive handshake packets, because it does consume and produce net data.

Important to note; handshakes can happen at any time.

##### Stream level

This appears attractive because it allows us to work with the `Stream` API. The major issue is that the two streams (input and output) are not independent. When doing the handshake, `SSLEngine` will request data from the input stream, and will request us to push data to the output stream.

So the output stream, which during normal operation would only deal with `wrap`, could run into a `NEED_UNWRAP` status, and it would need to get bytes from the input stream and feed them into `SSLEngine`, which is a problem, since it doesn't have access to the input stream.

##### Socket level

At the socket level, we have access to both input and output streams, so that problem is solved. I think at the socket level it's a pretty straightforward implementation.

One important thing is that unlike `Socket`, which never buffers bytes, something using `SSLEngine` will have to buffer bytes: `SSLEngine` will only decode full TLS packets; so if a client requests a single byte, we still need to read a whole TLS packet from the net, unwrap it, and buffer the remainder. 

Several variations are possible:

* Decorator for `Socket`. This is nice because it requires no changes to `Socket` or anything else in  `fs2.io.tcp`. A disadvantage is that `java.nio` and `SSLEngine` both work with `ByteBuffer`, while `Socket` uses `Chunk[Byte]`. If we decorate `Socket`, we'll have to go back and forth a lot.
* Alternative `Socket` implementation. To obtain an instance, we'll probably need new methods on `fs2.io.tcp` as well, `sslServer`, `sslClient`, `sslServerWithLocalAddress`, or add an `Option[() => SSLEngine] = None` parameter to the existing ones.

I'm curious to hear opinions :)

**Comment by pchlupacek** [comment]
@eamelink my thought would be to have something like 

```scala def ssl[F[_]](sslparams....)(socket:Socket[F]):Socket[F] = ???```

**Comment by pchlupacek** [comment]
@eamelink can you perhaps update how you are progressing? I may be in need of that functionality earlier than expected, so was perhaps hoping if we can join forces on this?

**Comment by dscleaver** [comment]
That reads almost exactly like my experience when I tried this before.

**Comment by eamelink** [comment]
I put an initial attempt at https://github.com/eamelink/fs2/tree/tls

There's a sample program included with an echo-server, which properly responds when the user disconnects, and also a server that doesn't read anything, but just outputs some chunks and then closes the connection from the server end. 

There are still many things buggy and not implemented.

My next steps are writing tests to verify correct behaviour, and when tests are in place, we can experiment with simpler designs as well.

**Comment by eamelink** [comment]
Today @pchlupacek has given a lot of advice and suggestions for much simpler concurrency control,  so there are a lot of improvements and simplifications coming up :)

**Comment by mpilquist** [comment]
@eamelink Any updates on this?

**Comment by eamelink** [comment]
Thanks for reminding me!

I pushed the latest status to https://github.com/eamelink/fs2/commit/e95ec6188603beffd1fb2e44d41a86bb828c7419

This works, and there is a sample server application included to demonstrate it. There's still a fair amount of loose ends though, written down in the [Readme](https://github.com/eamelink/fs2/blob/e95ec6188603beffd1fb2e44d41a86bb828c7419/io/src/main/scala/fs2/io/tcp/tls/README.md)

The next priority is testing, but I need some help with that. I'll ask the folks in the Gitter channel.

**Comment by eamelink** [comment]
Small update. I have some tests, which uncovered a fair amount of issues, some regarding proper closing when a connection stops for one of the various reasons, and harder ones regarding rehandshaking during active receiving and transmitting. But there is still progress :)

**Comment by pchlupacek** [comment]
I have completed implementation of this. So far it is here https://github.com/Spinoco/fs2-http/tree/master/src/main/scala/spinoco/fs2/interop/ssl, but would be extracting this against 1.0 branch soon.

**Comment by pchlupacek** [comment]
resolved in #804

**Comment by pchlupacek** [comment]
After discussing with @mpilquist we will not add this currently to fs2. There is however support for TLS available here: https://github.com/Spinoco/fs2-crypto
